File: leetcode/e/Pow(x, n).java
/**
 Implement pow(x, n).
 **/

public class Solution {
  // not pass edge case
  public double myPow(double x, int n) {
    if (n == 0) return 1;
    if (n < 0) return 1 / myPow(x, -1 * n); // bug: Must consider minus value and MIN_VALUE can overflow
    if (n % 2 == 0) return myPow(x, n / 2);
    else return x * myPow(x, (n - 1) / 2);
  }

  // correct
  public double myPow(double x, int n) {
    if (n == 0) return 1;
    double val = myPow(x, n / 2);
    if (n % 2 == 0)  return val * val;
    else {
      if (n > 0)
        return x * val * val;
      else return val * val / x;
    }
  }
}

File: lc_plusplus/Pow(x, n).cc
// Implement pow(x, n).
//
//
// Example 1:
//
// Input: 2.00000, 10
// Output: 1024.00000
// Example 2:
//
// Input: 2.10000, 3
// Output: 9.26100


class Solution {
public:
    double myPow(double x, int n) {
        if (n == 0) return 1;
        auto val = myPow(x, n / 2);
        if (n % 2 == 0) {
            return val * val;
        }
        if (n > 0) return x * val * val;
        else return val * val / x;
    }
};
